README.md
UnityEngine2deb
===============

Create Debian packages of Unity Engine games

**u2deb**: Package a Unity game in two steps. First run `u2deb prepare <gamedir>` and follow the instructions.
The game files will be copied into a temporary directory and Debian packaging files will be generated.
You can manually edit those files. Then run `u2deb build` to create the package. To delete the temporary files
run `u2deb clean`. You can get a full list of options and environment variables with `u2deb --help`.

**copyunityengine**: This script will download and install GNU/Linux binaries of a Unity engine game and delete all previously
installed binary files. Usage is `copyunityengine <gamedir>`.
You can use this script to create native GNU/Linux versions of games that were released only for
Windows and/or OSX, or to create native 64 bit versions if a GNU/Linux build was originally released only as 32 bit version.
You can also use it to install the screenselector plugin if the original build came without it.

**u2mojo**: This script helps you to create portable [MojoSetup](http://www.icculus.org/mojosetup/)
installers for any GNU/Linux distribution.

Minimum dependencies: `debhelper imagemagick aria2¹ p7zip-full¹ wget¹`

Recommended dependencies: `execstack lintian libgtk2.0-0 libgtk2.0-0:i386 libglu1-mesa libglu1-mesa:i386`

¹ only required by copyunityengine

**Examples:**

You can use the builds from [here](https://github.com/darealshinji/simple-unity3d-example/releases) to test the scripts.

Build 32 and 64 bit packages of a native game and store the platform independent
files in a separate "-data" package:
```
./u2deb prepare example-game-linux --data
./u2deb build
```

Build a 64 bit package from a 32 bit native game:
```
./copyunityengine example-game-linux-x86
./u2deb prepare example-game-linux-x86 --no-x86
./u2deb build
```

Build packages from a Windows-only game:
```
./copyunityengine example-game-win32
rm -f example-game-win32/*.exe example-game-win32/*.pdb example-game-win32/*_Data/Mono/mono.dll
./u2deb prepare example-game-win32
./u2deb build
```

Build packages from an OS X .app bundle:
```
./copyunityengine example-game.app
./u2deb prepare example-game-linux
./u2deb build
```

Build a MojoSetup installer and export variables:
```
echo "This is a simple Unity 3D test project." > /tmp/example-game-description
echo "Homepage: http://www.unity3d.com/" >> /tmp/example-game-description

export FULLNAME="Test project"
export SHORTNAME=example-game
export VENDOR="djcj"
export VERSION=1.0
./u2mojo --readme=/tmp/example-game-description example-game
```

**Package this tool:**

You can also build a Debian package from this tool.
```
sudo apt-get install debhelper git
git clone https://github.com/darealshinji/UnityEngine2deb.git
cd UnityEngine2deb
dpkg-buildpackage -b -us -uc
```


debian/changelog
u2deb (16.04.29.1~trusty) trusty; urgency=low

  * copyunityengine: support OSX .app bundles as input directories

 -- Marshall Banana <djcj@gmx.de>  Fri, 13 May 2016 07:22:15 +0200

u2deb (16.02.24.1~trusty) trusty; urgency=low

  * copyunityengine:
      * parse commands in a getopt()-like way
      * new option to create a tarball from a specified exe file
      * print md5 checksum after creating an archive
  * u2deb:
      * clarify that the UNITY-SOFTWARE-LICENSE-AGREEMENT applies to games
      * update PatchELF
  * u2mojo:
      * fix download of MojoSetup binaries and add MD5 check

 -- Marshall Banana <djcj@gmx.de>  Thu, 31 Mar 2016 16:45:53 +0200

u2deb (16.01.14.1~trusty) trusty; urgency=low

  * copyunityengine:
      * get version number from Unity 5 games
  * u2deb:
      * convert ScreenSelector.bmp to .png
      * delete ScreenSelector.bmp
      * automatic fix of DOS line endings in descriptions

 -- Marshall Banana <djcj@gmx.de>  Thu, 14 Jan 2016 13:44:44 +0100

u2deb (15.12.08.1~trusty) trusty; urgency=low

  * copyunityengine:
      * show confirmation dialog before downloading huge files
      * inform when the checksum is okay
      * enable automatic download for Unity >= 5.3 (files will be downloaded
        with Aria2c via BitTorrent)
  * New scripts:
      * u2mojo: create Mojo Setup packages of Unity engine games

 -- Marshall Banana <djcj@gmx.de>  Tue, 08 Dec 2015 23:10:12 +0100

u2deb (15.09.09.1~trusty) trusty; urgency=low

  * copyunityengine: fix checking for aria2c

 -- Marshall Banana <djcj@gmx.de>  Wed, 09 Sep 2015 07:31:52 +0200

u2deb (15.08.15.1~trusty) trusty; urgency=low

  * copyunityengine: small modification to enable downloads from Google Drive

 -- Marshall Banana <djcj@gmx.de>  Sat, 15 Aug 2015 04:44:55 +0200

u2deb (15.08.11.1~trusty) trusty; urgency=low

  * copyunityengine: fix copying plugins

 -- Marshall Banana <djcj@gmx.de>  Tue, 11 Aug 2015 22:44:10 +0200

u2deb (15.07.30.2~trusty) trusty; urgency=low

  * copyunityengine: fix downloading engine packages

 -- Marshall Banana <djcj@gmx.de>  Thu, 30 Jul 2015 15:31:00 +0200

u2deb (15.07.30.1~trusty) trusty; urgency=low

  * New scripts:
      * copyunityengine: download and install Linux binaries of a specific
        engine version
  * Create random build directory
  * Don't fail package building if there's a 64 bit binary but no 32 bit one
  * Ship with a copy of PatchELF so we don't have to rely on it as a
    third-party dependency
  * Install files into /usr/lib/games

 -- Marshall Banana <djcj@gmx.de>  Thu, 30 Jul 2015 13:37:34 +0200

u2deb (15.03.26.1~trusty) trusty; urgency=low

  * Copy only game_Data and game.x86*
  * Delete any *.dll.mdb file

 -- Marshall Banana <djcj@gmx.de>  Thu, 26 Mar 2015 17:31:33 +0100

u2deb (15.03.16.1~trusty) trusty; urgency=low

  * Add missing ScreenSelector.png icon if required
  * New default icon (vectorized Unity logo)

 -- Marshall Banana <djcj@gmx.de>  Mon, 16 Mar 2015 17:47:34 +0100

u2deb (15.03.13.1~trusty) trusty; urgency=low

  * Removing unused DT_NEEDED entries from ELF file headers can now be turned
    off with --no-patchelf
  * i386 packages created on amd64 architectures do no longer have a double
    libc6 dependency entry in their control file

 -- Marshall Banana <djcj@gmx.de>  Fri, 13 Mar 2015 13:07:37 +0100

u2deb (15.03.12.1~trusty) trusty; urgency=low

  * Enable -V and -h commands even if templates weren't found
  * Run execstack only if the command is available
  * Implement --no-x86 and --no-x86_64 commands to disable the build of
    i386 or amd64 packages
  * With --working-dir=<path> the temporary files can now be stored in
    a different location

 -- Marshall Banana <djcj@gmx.de>  Thu, 12 Mar 2015 20:20:45 +0100

u2deb (15.02.04.1~trusty) trusty; urgency=low

  * Delete .DS_Store and Thumbs.db files
  * Delete __MACOSX directories
  * Use PatchELFmod only
  * Add UPSTREAMNAME and FILENAME to environment variables
  * Move the following actions from debian/rules to the shell script:
    * remove executable bits
    * remove executable stack
    * delete unnecessary files
  * Remove Lintian overrides

 -- Marshall Banana <djcj@gmx.de>  Wed, 04 Feb 2015 16:00:34 +0100

u2deb (15.01.29.1~trusty) trusty; urgency=low

  * Automatically delete UnityEngine debug data
  * Find and delete Steam API plugins

 -- Marshall Banana <djcj@gmx.de>  Thu, 29 Jan 2015 18:49:47 +0100

u2deb (15.01.27.3~trusty) trusty; urgency=low

  * Don't build arch-independend *-data packages twice

 -- Marshall Banana <djcj@gmx.de>  Tue, 27 Jan 2015 17:30:40 +0100

u2deb (15.01.27.2~trusty) trusty; urgency=low

  * Implement --output/-o command line options
  * Override Lintian warnings about ScreenSelector.png in inside /usr/lib
  * Add Lintian to dependencies
  * Make Debian packages depend on libpulse0

 -- Marshall Banana <djcj@gmx.de>  Tue, 27 Jan 2015 11:23:45 +0100

u2deb (15.01.27.1~trusty) trusty; urgency=low

  * New version
  * Implement a non-interactive way to enter packaging information
    via the shell's export function
  * Fix templates/rules to not fail if the game has no plugin directory

 -- Marshall Banana <djcj@gmx.de>  Tue, 27 Jan 2015 08:26:59 +0100

u2deb (15.01.07.1~trusty) trusty; urgency=low

  * New version

 -- Marshall Banana <djcj@gmx.de>  Sat, 10 Jan 2015 12:15:08 +0100

u2deb (14.12.29.1~trusty1) trusty; urgency=low

  * New version

 -- Marshall Banana <djcj@gmx.de>  Sat, 03 Jan 2015 05:31:12 +0100

u2deb (14.10.09.1~trusty2) trusty; urgency=low

  * Override Lintian warnings

 -- Marshall Banana <djcj@gmx.de>  Tue, 11 Nov 2014 13:56:28 +0100

u2deb (14.10.09.1~trusty1) trusty; urgency=low

  * New version

 -- Marshall Banana <djcj@gmx.de>  Thu, 09 Oct 2014 00:27:55 +0200

u2deb (14.09.14.1~trusty2) trusty; urgency=low

  * Upload to PPA

 -- Marshall Banana <djcj@gmx.de>  Sun, 14 Sep 2014 20:01:35 +0200

u2deb (14.09.07.1~1) unstable; urgency=low

  * Initial release


templates/make-icons.sh
#!/bin/sh -e

# Create an X11 icon set from a single image file.
# Usage: make-icon.sh <icon> [<output-folder>]

# Last revision : 2014-09-07
# Requires: librsvg2-bin imagemagick


# Copyright (c) 2014, djcj <djcj@gmx.de>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.


input="$1"
output1=$(basename "$1")
output="${output1%.*}"
tmp="${output1}_tmp.png"

if [ -z "$1" ] ; then
    echo "Usage: $0 <icon> [<output-folder>]"
    exit 1
elif [ ! -e "$1" ] ; then
    echo "Can't find '$1'!"
    exit 1
elif [ -d "$1" ] ; then
    echo "Input is a directory!"
    exit 1
fi

if [ -z "$2" ] ; then
    destdir=converted_icons
else
    destdir="$2"
fi


## Check if input is an SVG
if [ "${input##*.}" = "svg" ] || [ "${input##*.}" = "SVG" ] ; then
    svgdir="$destdir/icons/hicolor/scalable/apps"
    echo "install SVG"

    # install SVG into hicolor/scalable/apps
    install -c -D -m644 "$input" "$svgdir/$output1"

    # Create temporary input PNG file
    rsvg-convert "$input" -o "$tmp"
    input="$tmp"
fi


## Create XPM
pixmaps="$destdir/pixmaps"
mkdir -p "$pixmaps"
echo "create pixmap"
convert "$input" -filter Lanczos -resize 32x32 "$pixmaps/$output.xpm"


## Create PNGs
for n in 16 22 24 32 48 64 96 128 256 512
do
    w=$(identify -format "%w" "$input")
    h=$(identify -format "%h" "$input")

    if [ $w != $h ] ; then
        echo "hicolor icons require width and height to be equal, but input resolution is ${w}x${h}."
        echo "Didn't create hicolor icons."
        exit 0
    fi

    if [ $w -ge $n ] ; then
        icondestdir="$destdir/icons/hicolor/${n}x${n}/apps"
        mkdir -p "$icondestdir"
        echo "create hicolor ${n}x${n}"
        convert "$input" -filter Lanczos -resize ${n}x${n} "$icondestdir/$output.png"
    fi
done

if [ -f "$tmp" ] ; then
    rm "$tmp"
fi


u2deb.sh
#!/bin/bash

set -e

# Copyright (c) 2014-2015, djcj <djcj@gmx.de>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

LANG=C
LANGUAGE=C
LC_ALL=C

appversion="15.06.23.1"

appname=$(basename "$0")
scriptpath="$(dirname "$(readlink -f "$0")")"
linkname="UnityEngine2deb_working_directory"
defaultwd="/tmp/UnityEngine2deb_tmp"

errorExit() {
  echo "error: $1"
  exit 1
}

help() {
cat << EOF

 Create Debian packages of Unity Engine games

 Usage:
   $appname -h|--help|-V|--version
   $appname -p|prepare <path> [OPTIONS]
   $appname -b|build|make [OPTIONS]
   $appname -c|clean [OPTIONS]

 Options:
   -h, --help           print this message
   -V, --version        display version info and quit

   -p, prepare <path>   copy files from <path> and prepare a Debian
                           source package
   -b, build, make      build binary packages
   -c, clean            delete the working tree

   -o=<path>,
   --output=<path>      save Debian packages in <path>
   --working-dir=<path> Working directory where the temporary files are stored.
                           Default: $defaultwd

   -d, --data           build a separate package for architecture-
                           independent files
   --gzip               use gzip instead of xz compression (faster compression
                           at the cost of bigger packages)
   --icon=<icon>        use this icon for the desktop entry
   --no-x86             don't build an i386 package
   --no-x86_64          don't build an amd64 package
   --no-patchelf        don't remove unused DT_NEEDED entries from ELF binary
                           file headers

 Environment variables:
   UPSTREAMNAME         the original name of the game, including special chars
                           or spaces
   PKGNAME              specify a name for the executable and the package
   SHORTDESCRIPTION     a brief game description for the package and menu entry
   VERSION              the game's upstream version
   MAINTAINER           The package maintainer. Make sure to use the following
                           pattern: John Doe <nick@domain.org>
   HOMEPAGE             homepage of the game or the developer
   YEAR                 the year when the game was released
   COPYRIGHT            Who's holding the copyright?

EOF
}

if [ "$1" = "-V" ] || [ "$1" = "--version" ] ; then
  echo $appversion
  exit 0
fi
if [ -z "$1" ] || [ "$1" = "-h" ] || [ "$1" = "help" ] ; then
  help
  exit 0
fi

if [ -d "$scriptpath/templates" ] ; then
  templates="$scriptpath/templates"
elif [ -d "$scriptpath/../share/$appname" ] ; then
  templates="$scriptpath/../share/$appname"
elif [ -d "/usr/local/share/$appname" ] ; then
  templates="/usr/local/share/$appname"
elif [ -d "/usr/share/$appname" ] ; then
  templates="/usr/share/$appname"
else
  errorExit "Can't find the templates!"
fi


datapackage="no"
icon=""
mode="empty"
output="$HOME"
compression="xz"
disable_x86="no"
disable_x86_64="no"
patchelfmod="yes"
wd=""
for opt; do
  optarg="${opt#*=}"
  case "$opt" in
    "prepare"|"-p")
      mode="prepare"
      ;;
    "build"|"-b"|"make")
      mode="build"
      ;;
    "clean"|"-c")
      mode="clean"
      ;;
    --output=*|-o=*)
      output="$optarg"
      ;;
    --working-dir=*)
      wd="$optarg"
      ;;
    "--data"|"-d")
      datapackage="yes"
      ;;
    --gzip)
      compression="gzip"
      ;;
    --icon=*)
      icon="$optarg"
      ;;
    "--no-x86")
      disable_x86="yes"
      ;;
    "--no-x86_64")
      disable_x86_64="yes"
      ;;
    "--no-patchelf")
      patchelfmod="no"
      ;;
    "help"|"-h"|"--version"|"-V")
      ;;
    *)
      origpath="$optarg"
      ;;
  esac
done

if [ $disable_x86 = "yes" ] && [ $disable_x86_64 = "yes" ] ; then
  errorExit "you can't use \`--no-x86' together with \`--no-x86_64'"
fi

if [ $mode = "empty" ] ; then
  help
  exit 1
fi

if ([ $mode = "build" ] || [ $mode = "clean" ]) && [ -L "$PWD/$linkname" ] && [ -z "$wd" ] ; then
  wd="$(readlink "$PWD/$linkname")"
fi
[ -n "$wd" ] && topsrc="$wd" || topsrc="$defaultwd"
builddir="$topsrc/build"
sourcedir="$topsrc/source"
icondir="$builddir/icon"
debian="$builddir/x86/debian"



################## clean ##################
if [ $mode = "clean" ] ; then
  rm -rvf "$topsrc"
  [ ! -L "$PWD/$linkname" ] || rm -v "$PWD/$linkname"
  exit 0
fi


################# prepare #################
if [ $mode = "prepare" ] ; then
  if [ -z "$origpath" ] ; then
    errorExit "no path specified"
  elif [ ! -e "$origpath" ] ; then
    errorExit "path to '$origpath' doesn't exist"
  elif [ ! -d "$origpath" ] ; then
    errorExit "'$origpath' is not a directory"
  else
    path="$( cd "$origpath" && pwd )"
  fi
  rm -rf $cleanfiles

  # get the application name
  filename="$(basename "$(find "$origpath" -type d -name *_Data)" | head -c-6)"
  [ -z "$PKGNAME" ] && name="$filename" || name="$PKGNAME"
  [ -z "$UPSTREAMNAME" ] && UPSTREAMNAME="$name"
  if [ -n "$UPSTREAMNAME" ] && [ -n "$PKGNAME" ] ; then
    name="$PKGNAME"
  fi

  # check for architectures
  x86="no"
  x86_64="no"
  rename_to_x86="no"
  rename_to_x86_64="no"
  [ -f "$origpath/$filename".x86 ] && x86="yes"
  [ -f "$origpath/$filename".x86_64 ] && x86_64="yes"
  if [ $x86 = "no" ] && [ $x86_64 = "no" ] ; then
    echo "neither '$filename.x86' nor '$filename.x86_64' found"
    if [ -f "$origpath/$filename" ] ; then
      if [ "$(file "$origpath/$filename" | grep 'ELF 32-bit')" ]; then
        echo "'$filename' (x86) found"
        x86="yes"
        rename_to_x86="yes"
      elif [ "$(file "$origpath/$filename" | grep 'ELF 64-bit')" ]; then
        echo "'$filename' (x86_64) found"
        x86_64="yes"
        rename_to_x86_64="yes"
      else
        errorExit "'$filename' was found but is not a valid ELF binary"
      fi
    else
      errorExit "couldn't find '$filename' either"
    fi
  fi
  if [ $x86_64 = "yes" ] && [ $x86 = "no" ]; then
    disable_x86="yes"
  fi
  if [ $disable_x86 = "yes" ] && [ $x86_64 = "no" ] ; then
    errorExit "x86 disabled but no x86_64 binary found"
  elif [ $disable_x86_64 = "yes" ] && [ $x86 = "no" ] ; then
    errorExit "x86_64 disabled but no x86 binary found"
  fi

  # create working directory and symlink
  [ ! -L "$PWD/UnityEngine2deb_working_directory" ] || rm "$PWD/UnityEngine2deb_working_directory"
  rm -rf "$topsrc"
  mkdir -p "$topsrc"
  ln -s "$topsrc" "$PWD/UnityEngine2deb_working_directory"

  # copy source files
  echo "copy source files... "
  mkdir -p "$sourcedir"
  cp -vr "$origpath"/* "$sourcedir"
  echo "done"

  # remove executable bits
  find "$sourcedir" -type f -exec chmod a-x '{}' \;

  # remove executable stack
  export PATH="/usr/local/sbin:/usr/sbin:/sbin:${PATH}"  # on Debian /sbin isn't in PATH by default
  set +e; which execstack 2>/dev/null >/dev/null; set -e
  exitCode=$(echo $?)
  if [ $exitCode = 0 ]; then
    find "$sourcedir" -name libmono.so -exec execstack -c '{}' \;
  fi

  # delete unnecessary files
  for f in libCSteamworks.so libsteam_api.so libSteamworksNative.so SteamworksNative.dll \
           \*.dll.mdb Thumbs.db .DS_Store ;
  do
    find "$sourcedir" -name "$f" -delete
  done
  rm -rf `find "$sourcedir" -type d -name __MACOSX`

  # rename application
  name=$(echo "$name" | sed -e 's/\(.*\)/\L\1/; s/\ -\ /-/g; s/\ /-/g; s/_/-/g')
  [ "$filename" != "$name" ] && rename "s/$filename/$name/" "$sourcedir"/*
  if [ "$rename_to_x86" = "yes" ] && [ -f "$sourcedir/$name" ] ; then
    mv "$sourcedir/$name" "$sourcedir/${name}.x86"
    echo "application was renamed to ${name}.x86"
  elif [ "$rename_to_x86_64" = "yes" ] && [ -f "$sourcedir/$name" ] ; then
    mv "$sourcedir/$name" "$sourcedir/${name}.x86_64"
    echo "application was renamed to ${name}.x86_64"
  fi

  # add a ScreenSelector.png if required
  if [ -f "`find "$sourcedir" -name ScreenSelector.so | head -n1`" ] && \
     [ ! -f "$sourcedir/${name}_Data/ScreenSelector.png" ] ; then
    cp "$templates/ScreenSelector.png" "$sourcedir/${name}_Data/ScreenSelector.png"
  fi

  # icon
  default_icon="no"
  mkdir -p "$icondir"
  if [ -z "$icon" ] || [ ! -f "$icon" ] ; then
    if [ -f "$sourcedir/${name}_Data/Resources/UnityPlayer.png" ] ; then
      cp "$sourcedir/${name}_Data/Resources/UnityPlayer.png" "$icondir/$name.png"
      icon="$icondir/$name.png"
    else
      cp "$templates/icon.png" "$icondir/$name.png"
      icon="$icondir/$name.png"
      default_icon="yes"
    fi
    else
      cp "$icon" "$icondir/$name.png"
      icon="$icondir/$name.png"
  fi

  # enter packaging information
  echo ""
  echo "name: $UPSTREAMNAME"
  echo "application/package name: $name"
  if [ -z "$SHORTDESCRIPTION" ] ; then
    echo ""
    echo "Please enter a brief description,"
    read -p "i.e. 'Unity engine video game': " SHORTDESCRIPTION
  else
    echo "description: $SHORTDESCRIPTION"
  fi
  if [ -f "$PWD/description" ] ; then
    cp "$PWD/description" "$topsrc"
  else
    cp "$templates/description" "$topsrc"
    echo ""
    echo "Please add a more detailed description about the game in the text file"
    echo "$topsrc/description"
    echo "and press any key to continue."
    read -p "" -n1 -s
  fi
  echo "" >> "$topsrc/description"  # add a new line to the end of the file
  if [ -z "$VERSION" ] ; then
    echo ""
    echo "Enter the game's release version. It should begin with a number"
    read -p "and mustn't contain and spaces or underscores: " VERSION
  else
    echo "package version: $VERSION"
  fi
  if [ -z "$MAINTAINER" ] ; then
    echo ""
    echo "Enter the package maintainer information."
    echo "Use the following pattern: John Doe <nick@domain.org>"
    read -p " " MAINTAINER
  else
    echo "package maintainer: $MAINTAINER"
  fi
  if [ -z "$HOMEPAGE" ] ; then
    echo ""
    read -p "What's the game's homepage? " HOMEPAGE
  else
    echo "homepage: $HOMEPAGE"
  fi
  if [ -z "$YEAR" ] ; then
    echo ""
    read -p "What year is this game from? " YEAR
  else
    echo "year: $YEAR"
  fi
  if [ -z "$COPYRIGHT" ] ; then
    echo ""
    read -p "Who's holding the copyright? " COPYRIGHT
  else
    echo "copyright: $COPYRIGHT"
  fi
  echo ""
  echo ""
  [ -z "$SHORTDESCRIPTION" ] && SHORTDESCRIPTION="Unity engine video game"
  [ -z "$VERSION" ] && VERSION=$(date +%y.%m.%d.1)
  [ -z "$MAINTAINER" ] && MAINTAINER="John Doe <nick@domain.org>"
  [ -z "$HOMEPAGE" ] && HOMEPAGE="http://www.unity3d.com/"
  [ -z "$YEAR" ] && YEAR=$(date +%Y)
  [ -z "$COPYRIGHT" ] && COPYRIGHT="the creator of '$name'"

  # create Debian files
  mkdir -p "$debian"
  cp -r \
  "$templates/source" \
  "$templates/compat" \
  "$templates/lintian-overrides" \
  "$templates/rules" \
  "$templates/make-icons.sh" \
  "$templates/patchelfmod" "$debian"
  chmod a+x "$debian/rules" "$debian/make-icons.sh"

  cat >> "$builddir/x86/${name}.desktop" << EOF
[Desktop Entry]
Name=$UPSTREAMNAME
Comment=$SHORTDESCRIPTION
Exec=$name
Type=Application
Categories=Game;
StartupNotify=true
Icon=$name
EOF

  echo "debian/${name}.6" > "$builddir/x86/debian/${name}.manpages"

  cat >> "$builddir/x86/debian/${name}.6" << EOF
.TH ${name^^} 6 "" "$(date +%d.%m.%Y)"
.SH NAME
$name \- $SHORTDESCRIPTION
.SH SYNOPSIS
.B $name
.SH OPTIONS
This game has no command line options.
.SH DESCRIPTION
EOF
  fold -s "$topsrc/description" >> "$builddir/x86/debian/${name}.6"
  cat >> "$builddir/x86/debian/${name}.6" << EOF
.SH SEE ALSO
.I $HOMEPAGE
.SH AUTHOR
$COPYRIGHT
EOF

  cat >> "$debian/copyright" << EOF
Format: http://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
Upstream-Name: $UPSTREAMNAME
Source: $HOMEPAGE

Files: *
Copyright: $YEAR $COPYRIGHT
License:
 Copyright (c) $YEAR, $COPYRIGHT
 All Rights Reserved.

EOF
  cat "$templates/debian-copyright" >> "$debian/copyright"

  [ $datapackage = "yes" ] && datadeps=", ${name}-data (= \${binary:Version})"
  cat >> "$debian/control.in" << EOF
Source: $name
Section: games
Priority: optional
Maintainer: $MAINTAINER
Build-Depends: debhelper (>= 9)
Standards-Version: 3.9.5
Homepage: $HOMEPAGE

Package: $name
Architecture: any
Depends: \${misc:Depends}, \${shlibs:Depends}, libpulse0$datadeps
Description: $SHORTDESCRIPTION
EOF
  printf '%s\n' "`cat "$topsrc/description"`" | fold -s -w 79 | sed 's/^/ /g; s/^ $/ ./g; s/ $//g' >> "$debian/control.in"

  if [ $datapackage = "yes" ] ; then
    cat >> "$debian/control.in" << EOF

Package: ${name}-data
Architecture: all
Depends: \${misc:Depends}
Recommends: $name
Description: $SHORTDESCRIPTION - game data
EOF
    printf '%s\n' "`cat "$topsrc/description"`" | fold -s -w 79 | sed 's/^/ /g; s/^ $/ ./g; s/ $//g' >> "$debian/control.in"
    echo " ." >> "$debian/control.in"
    echo " This package installs the $name game data." >> "$debian/control.in"
  fi

  uniq "$debian/control.in" "$debian/control"

  cat >> "$debian/changelog" << EOF
$name ($VERSION) unstable; urgency=medium

  * Initial release
 
 -- $MAINTAINER  $(date -R)
EOF

  if [ $x86_64 = "yes" ] ; then
    mkdir -p "$builddir/x86_64/debian"
    cp -rf "$debian"/* "$builddir/x86_64/debian"
    cp -f "$builddir/x86/${name}.desktop" "$builddir/x86_64"
    cat >> "$builddir/x86_64/debian/confflags" << EOF
NAME = $name
ICON = "$icon"
ARCH = x86_64
DATAPACKAGE = $datapackage
Z = $compression
PATCHELFMOD = $patchelfmod
EOF
  fi
  cat >> "$debian/confflags" << EOF
NAME = $name
ICON = "$icon"
ARCH = x86
DATAPACKAGE = $datapackage
Z = $compression
PATCHELFMOD = $patchelfmod
EOF

  [ $disable_x86 = "yes" ] && rm -rf "$builddir/x86"
  [ $disable_x86_64 = "yes" ] && rm -rf "$builddir/x86_64"
  if [ -d "$builddir/x86" ] ; then
    echo "hard link files to build/x86... "
    mkdir -p "$builddir/x86/source"
    cp -vrl "$sourcedir/${name}_Data" "$sourcedir/${name}.x86" "$builddir/x86/source"
    echo "done"
  fi
  if [ -d "$builddir/x86_64" ] ; then
    echo "hard link files to build/x86_64..."
    mkdir -p "$builddir/x86_64/source"
    cp -vrl "$sourcedir/${name}_Data" "$sourcedir/${name}.x86_64" "$builddir/x86_64/source"
    echo "done"
  fi
  if [ -d "$builddir/x86" ] && [ -d "$builddir/x86_64" ] ; then
    echo "arch_only = --arch" >> "$builddir/x86_64/debian/confflags"
  fi
  echo ""
  echo "The files are prepared. Run '$appname build' to build Debian packages."
  echo "Or modify the the files in '$builddir' first."

  exit 0
fi


################## build ##################
buildpackage () {
  arch=${1}
  if [ -d "$builddir/${arch}" ] ; then
    name="$(grep 'NAME = ' $builddir/${arch}/debian/confflags | sed -e 's/NAME = //')"
    cd "$builddir/${arch}"
    chmod a+x debian/rules source/${name}.${arch}
    mv source/${name}.${arch} source/${name}
    test ${arch} = "x86" && purge=x86_64 || purge=x86
    rm -rf source/${name}_Data/Mono/${purge} source/${name}_Data/Plugins/${purge} source/${name}.${purge}
    dpkg-buildpackage -b -us -uc 2>&1 | tee "$builddir/${arch}-build.log"
  fi
}

if [ $mode = "build" ] ; then
  if [ ! -d "$builddir/x86" ] && [ ! -d "$builddir/x86_64" ] ; then
    echo "no files in '$builddir'!"
    errorExit "run '$appname prepare <path>' first"
  fi

  buildpackage x86
  buildpackage x86_64

  cd "$builddir"
  echo ""
  ls *.deb >/dev/null
  if [ $(echo $?) = 0 ] ; then
    for f in *.deb ; do
      echo "$f:"
      dpkg-deb -I $f
      lintian $f
      echo ""
    done 2>&1 | tee "$builddir/packages.log"
    for f in *.deb ; do
      echo "$f:"
      dpkg-deb -c $f
      echo ""
    done 2>&1 | tee -a "$builddir/packages.log"
  fi
  cp -f *.deb "$output"
  echo "Debian packages have been copied to '$output'"
fi

exit 0
